File: go/analyze/rules_rollback.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// -- Rule: require-down-migration -----------------------------------------

// RequireDownMigration warns when a versioned MigrationBundle does not
// have a downSource configured. Without a down source, the only
// recovery option on failure is a forward-fix — no automatic rollback
// is possible.
//
// Severity is warning (not error) because forward-fix is a legitimate
// strategy. Operators can suppress this per SchemaPolicy when they
// decide forward-fix is sufficient for their risk tier.
type RequireDownMigration struct{}

func (RequireDownMigration) ID() string { return "require-down-migration" }
func (RequireDownMigration) Description() string {
	return "warn when a versioned migration has no down source for rollback"
}

func (a *RequireDownMigration) Check(_ context.Context, m *Migration) ([]Finding, error) {
	// Only applies to versioned strategy — pgroll-expand-contract has
	// built-in abort handlers, and declarative diffs auto-generate
	// reverse SQL.
	if m.Strategy != "" && m.Strategy != string(keystonev1alpha1.StrategyVersioned) {
		return nil, nil
	}

	// Skip if no SQL files (operation-based bundle).
	if len(m.Files) == 0 {
		return nil, nil
	}

	if m.HasDownSource {
		return nil, nil
	}

	return []Finding{{
		Rule:     a.ID(),
		Severity: keystonev1alpha1.LintLevelWarning,
		Message: "no down source configured; if this migration fails, automatic rollback " +
			"is not possible — only forward-fix. Add spec.downSource with *.down.sql files " +
			"to enable rollback capability",
	}}, nil
}

File: rust/keystone-sdk/src/analyze/rules_rollback.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase 9.3 — rollback safety. Ported from `rules_rollback.go`.

use super::{finding, Analyzer, Finding, LintLevel, Migration};

/// require-down-migration
pub struct RequireDownMigration;
impl Analyzer for RequireDownMigration {
    fn id(&self) -> &'static str {
        "require-down-migration"
    }
    fn description(&self) -> &'static str {
        "warn when a versioned migration has no down source for rollback"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        // Only applies to the versioned strategy.
        if !m.strategy.is_empty() && m.strategy != "versioned" {
            return Vec::new();
        }
        // Skip operation-based bundles (no SQL files).
        if m.files.is_empty() {
            return Vec::new();
        }
        if m.has_down_source {
            return Vec::new();
        }
        vec![finding(
            self.id(),
            LintLevel::Warning,
            "",
            0,
            "no down source configured; if this migration fails, automatic rollback \
             is not possible — only forward-fix. Add spec.downSource with *.down.sql files \
             to enable rollback capability",
        )]
    }
}

#[cfg(test)]
mod tests {
    use super::super::{FileBody, Migration};
    use super::*;

    fn mig() -> Migration {
        Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: "CREATE TABLE t (id int);".into(),
            }],
            ..Default::default()
        }
    }

    #[test]
    fn fires_without_down_source() {
        assert_eq!(RequireDownMigration.check(&mig()).len(), 1);
    }

    #[test]
    fn silent_with_down_source() {
        let mut m = mig();
        m.has_down_source = true;
        assert_eq!(RequireDownMigration.check(&m).len(), 0);
    }

    #[test]
    fn skips_non_versioned_strategy() {
        let mut m = mig();
        m.strategy = "pgroll-expand-contract".into();
        assert_eq!(RequireDownMigration.check(&m).len(), 0);
    }

    #[test]
    fn skips_operation_based() {
        let mut m = mig();
        m.files.clear();
        assert_eq!(RequireDownMigration.check(&m).len(), 0);
    }
}
